ecc/goldilocks/decaf.go

```diff
@@ -0,0 +1,137 @@
+func (d Decaf) Neg(c, a *Elt) { c.p = a.p; c.p.cneg(1) }
+
+// Mul calculates c=n*a, where * is scalar multiplication on the group.
+func (d Decaf) Mul(c *Elt, n *Scalar, a *Elt) { d.c.ScalarMult(&c.p, n, &a.p) }
+
+// MulGen calculates c=n*g, where * is scalar multiplication on the group,
+// and g is the generator of the group.
+func (d Decaf) MulGen(c *Elt, n *Scalar) { d.c.ScalarBaseMult(&c.p, n) }
+
+// IsIdentity returns True if e is the identity of the group.
+func (e *Elt) IsIdentity() bool { return fp.IsZero(&e.p.x) }
+
+// IsEqual returns True if e=a, where = is an equivalence relation.
+func (e *Elt) IsEqual(a *Elt) bool {
+	l, r := &fp.Elt{}, &fp.Elt{}
+	fp.Mul(l, &e.p.x, &a.p.y)
+	fp.Mul(r, &a.p.x, &e.p.y)
+	fp.Sub(l, l, r)
+	return fp.IsZero(l)
+}
+
+// UnmarshalBinary if succeeds returns a Decaf element by decoding the first
+// DecafEncodingSize bytes of b.
+func (e *Elt) UnmarshalBinary(b []byte) error {
+	if len(b) < DecafEncodingSize {
+		return errInvalidDecoding
+	}
+
+	s := &fp.Elt{}
+	copy(s[:], b[:DecafEncodingSize])
+	isNeg := fp.Parity(s)
+	modulus := fp.P()
+	if isNeg == 1 || !isLessThan(b[:DecafEncodingSize], modulus[:]) {
+		return errInvalidDecoding
+	}
+
+	one, s2, den, num := &fp.Elt{}, &fp.Elt{}, &fp.Elt{}, &fp.Elt{}
+	isr, altx := &fp.Elt{}, &fp.Elt{}
+	t0, t1 := &fp.Elt{}, &fp.Elt{}
+	fp.SetOne(one)
+	fp.Sqr(s2, s)                    // s2  = s^2
+	fp.Sub(den, one, s2)             // den = 1 + a*s^2
+	fp.Mul(t1, s2, &paramDTwist)     // t1 = d*s^2
+	fp.Add(t1, t1, t1)               // t1 = 2*d*s^2
+	fp.Add(t1, t1, t1)               // t1 = 4*d*s^2
+	fp.Sqr(t0, den)                  // num = (1 + a*s^2)^2
+	fp.Sub(num, t0, t1)              // num = (1 + a*s^2)^2 - 4*d*s^2
+	fp.Mul(t0, t0, num)              // t0 = num*den^2
+	isQR := fp.InvSqrt(isr, one, t0) // v = 1/sqrt(num*den^2)
+	if !isQR {
+		return errInvalidDecoding
+	}
+	fp.Mul(t1, den, isr)                // altx = isr*den
+	fp.Mul(t1, t1, s)                   // altx = s*isr*den
+	fp.Add(t1, t1, t1)                  // t1   = 2*s*isr*den
+	fp.Mul(altx, t1, &sqrtAMinusDTwist) // altx = 2*s*isr*den*sqrt(A-D)
+	isNeg = fp.Parity(altx)             // isNeg = sgn(altx)
+	fp.Neg(t0, isr)                     // t0 = -isr
+	fp.Cmov(isr, t0, uint(isNeg))       // if altx is negative then isr = -isr
+	fp.Sqr(&e.p.x, isr)                 // x = isr^2
+	fp.Mul(&e.p.x, &e.p.x, den)         // x = isr^2*den
+	fp.Mul(&e.p.x, &e.p.x, num)         // x = isr^2*den*num
+	fp.Mul(&e.p.x, &e.p.x, s)           // x = s*isr^2*den*num
+	fp.Add(&e.p.x, &e.p.x, &e.p.x)      // x = 2*s*isr^2*den*num
+	fp.Mul(&e.p.y, isr, den)            // y = isr*den
+	fp.Add(t0, one, s2)                 // t0 = 1 - a*s^2
+	fp.Mul(&e.p.y, &e.p.y, t0)          // y = (1 - a*s^2)*isr*den
+	e.p.ta, e.p.tb = e.p.x, e.p.y       // T = Ta*Tb = x*y
+	fp.SetOne(&e.p.z)
+	return nil
+}
+
+// MarshalBinary returns a unique encoding of the element e.
+func (e *Elt) MarshalBinary() ([]byte, error) {
```

Comment:
I wouldn't name these functions related to "Binary", since I would interpret the inputs/returns to be bits. I would prefer something related to "Bytes" or "Octets".

Also, this function take a mutable buffer as input and then write the output to that buffer, rather than always creating a new buffer.

Reply:
`MarhshalBinary` is the standard Go name for such a function: https://golang.org/pkg/encoding/#BinaryMarshaler

Reply:
You want to implement it so that an element can be automatically serialised by the `json` package or any other encoder/decoder.

Reply:
> `MarhshalBinary` is the standard Go name for such a function: https://golang.org/pkg/encoding/#BinaryMarshaler

TIL